src/mono/mono/metadata/marshal-lightweight.c

```diff
@@ -2096,10 +2096,13 @@ emit_delegate_invoke_internal_ilgen (MonoMethodBuilder *mb, MonoMethodSignature
 				mono_mb_emit_op (mb, CEE_CALL, target_method);
 			} else {
 				mono_mb_emit_ldarg (mb, 1);
-				mono_mb_emit_op (mb, CEE_CASTCLASS, target_class);
-				for (i = 1; i < sig->param_count; ++i)
-					mono_mb_emit_ldarg (mb, i + 1);
-				mono_mb_emit_op (mb, CEE_CALLVIRT, target_method);
+				for (i = 1; i <= sig->param_count; ++i)
+					mono_mb_emit_ldarg (mb, i);
+				mono_mb_emit_ldarg (mb, 0);
+				mono_mb_emit_ldflda (mb, MONO_STRUCT_OFFSET (MonoDelegate, method));
+				mono_mb_emit_byte (mb, CEE_LDIND_I);
+				mono_mb_emit_icall (mb, mono_get_addr_compiled_method);
```

Comment:
Two thoughts:

1. Why do `delegate->target` in IL and not in the code of `mono_get_addr_compiled_method` (or a new icall that takes `MonoObject* target, MonoDelegate *delegate`
2. I don't actually see how this passes `target` to the icall. is that `mono_mb_emit_ldarg (mb, 1)`  above?  but then aren't all the other method arguments in the way? am I reading  this wrong?
3.

Reply:
1. By this we should not depend on the target method to also be able to generate a wrapper for AOT which does not depend on it
2. You are right, `target` is passed to the icall by `mono_mb_emit_ldarg (mb, 1)`. I should change the order of pushing the arguments onto the stack. Firstly I should push the arguments which will be used by `calli` instruction (`for loop`)

Reply:
> By this we should not depend on the target method to also be able to generate a wrapper for AOT which does not depend on it

Sorry, I wrote the wrong thing. I meant instead of `delegate->method` in IR, do it in C.

what you have right now is emitting:

```c
  mono_get_addr_compiled_method (object, delegate->method)
```
and then in the body of `mono_get_addr_compiled_method(MonoObject *object, MonoMethod *method`

and what I'm suggesting is instead emit:

```c
  mono_get_addr_compiled_method(object, delegate);
```

and then in the body of `mono_get_addr_compiled_method(MonoObject *object, MonoDelegate *delegate)` you first do
```c
   MonoMethod *method = delegate->method;
```
and then continue as before. 

What you have is ok, I just wonder why we need to extract the method from the delegate in IR

Reply:
Now I understand and I agree with you. This will make the IL code simpler, I will update it.